### DISCUSSION TITLE
The project descriptions says "low input lag"--what are the details?

### DISCUSSION THREAD

**Discussion by ChocoChopin** [ndeadly/MissionControl#489] [Q&A] [answered]
I'm assuming there isn't _zero_ input lag. Does anybody know just how much there is? Does it vary by input device? Does having the Mission Control service enabled create any extra input lag for native Switch controllers? If so, does disabling the service revert that issue?

**Comment by ndeadly** [comment, marked as answer]
I've never actually done any quantitative testing. The dot point is mainly there because many people seem to believe that Bluetooth inherently has large input delays, and that this will make them worse. Supposedly, on the Switch USB controllers actually have worse input delay than Bluetooth.

Of course, by introducing an additional link in the chain of processing input data, Mission Control increases the input delay on all controllers it handles (including official controllers) _in theory_. The amount of possible delay can vary between controllers depending on how many CPU instructions a controller handler needs to convert one input format to another. E.g controllers that support motion controls also have to do a bunch of additional calculations to calibrate the motion data, whereas official controllers just copy over their data more or less untouched. In practice though, I don't think these delays are significant enough to have any tangible impact on the existing delays without mission control.

If someone had the time and means to conduct some accurate tests I would love to see the results.

**Reply by ChocoChopin** [reply]
Wow, I'm amazed by the swiftness and comprehensiveness of your answer, thank you.

While I don't have a dedicated slowmo camera, I might use my iPhone to film at 240fps and see if I can detect any kind of lag at that framerate, doing a comparison with sysNAND.

**Reply by DARKFiB3R** [reply]
Did you ever perform any tests? I would be very interested in the results.

Thanks.

**Reply by ndeadly** [reply]
> Did you ever perform any tests? I would be very interested in the results.
> 
> Thanks.

I designed a setup and started doing some preliminary tests a couple of months ago. I haven't had the time to complete everything, but these are the current standings:

![current_standings-1.PNG](https://github.com/ndeadly/MissionControl/assets/24677491/90dc2119-ea71-4a79-b831-928dfc52b5f0)

Basically, we can conclude so far:
- Any delay introduced by Mission Control is negligible
- Bluetooth outperforms USB on the Switch
- Some Bluetooth controllers (such as DS4 seen above) can actually outperform native Switch controllers
- Wireless USB dongles suffer from the worst delay, due to having both Bluetooth and USB in the signal chain
- (Not shown here) there is an additional ~70ms delay between the application receiving an input and a resulting change being rendered to the screen that dwarves the actual controller delay
